File: lachancuocgoi/app/src/main/java/com/example/lachancuocgoi/Analysis/L3/GeminiSummarizer.kt
package com.example.lachancuocgoi.Analysis.L3

import android.util.Log
import com.example.lachancuocgoi.Analysis.L3.core.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class GeminiSummarizer(
    private val apiKeyProvider: ApiKeyProvider = BuildConfigApiKeyProvider(),
    private val keyHealthTracker: KeyHealthTracker? = null
) {
    
    companion object {
        private const val TAG = "GeminiSummarizer"
        private const val MIN_WORDS = 5
    }
    
    // ============ NEW: Sử dụng GeminiClient thống nhất ============
    private val geminiClient = GeminiClient(
        apiKeyProvider = apiKeyProvider,
        config = GeminiConfig.forSummarization(),
        keyHealthTracker = keyHealthTracker
    )
    
    /**
     * Tóm tắt đoạn hội thoại thành 1 câu ngắn gọn.
     * 
     * @param text Nội dung cuộc gọi cần tóm tắt
     * @return Tóm tắt 1 câu hoặc error message
     */
    suspend fun summarize(text: String): String = withContext(Dispatchers.IO) {
        // Validation
        val trimmed = text.trim()
        val wordCount = trimmed.split(Regex("\\s+")).size
        
        if (wordCount < MIN_WORDS) {
            return@withContext "Không đủ nội dung để tóm tắt (cần ít nhất $MIN_WORDS từ)"
        }
        
        // Create prompt
        val prompt = createPrompt(trimmed)
        
        // Query Gemini thông qua unified client
        val result = geminiClient.query(prompt) { responseText, _ ->
            responseText.trim()  // Parser đơn giản - chỉ trim whitespace
        }
        
        // Handle result
        return@withContext result.getOrElse { error ->
            Log.e(TAG, "Summarization failed: ${error.message}", error)
            "Lỗi tóm tắt: ${error.message}"
        }
    }
    
    /**
     * Tạo prompt cho summarization task.
     */
    private fun createPrompt(text: String): String {
        return PromptBuilder.buildSummarizationPrompt(text)
    }
}

File: lachancuoicgoi_flutter/lib/analysis/l3/gemini_summarizer.dart
import 'core/api_key_provider.dart';
import 'core/gemini_client.dart';
import 'core/gemini_config.dart';
import 'core/key_health_tracker.dart';
import 'prompt_builder.dart';
import '../../data/cloud_analysis_consent_store.dart';

class GeminiSummarizer {
  GeminiSummarizer({
    ApiKeyProvider? apiKeyProvider,
    KeyHealthTracker? keyHealthTracker,
    GeminiClient? geminiClient,
    CloudAnalysisConsentStore? cloudConsentStore,
  }) : _apiKeyProvider = apiKeyProvider ?? EnvironmentApiKeyProvider(),
       _keyHealthTracker = keyHealthTracker,
       _geminiClient = geminiClient,
       _cloudConsentStore = cloudConsentStore;

  static const int _minWords = 5;

  final ApiKeyProvider _apiKeyProvider;
  final KeyHealthTracker? _keyHealthTracker;
  final GeminiClient? _geminiClient;
  final CloudAnalysisConsentStore? _cloudConsentStore;
  GeminiClient? _cachedClient; // Cache client để preserve circuit breaker state

  GeminiClient get _client {
    if (_geminiClient != null) return _geminiClient;
    return _cachedClient ??= GeminiClient(
      apiKeyProvider: _apiKeyProvider,
      config: GeminiConfig.forSummarization(),
      keyHealthTracker: _keyHealthTracker,
      cloudConsentStore: _cloudConsentStore,
    );
  }

  Future<String> summarize(String text) async {
    final trimmed = text.trim();
    final words = trimmed
        .split(RegExp(r'\s+'))
        .where((part) => part.isNotEmpty);
    if (words.length < _minWords) {
      return 'Không đủ nội dung để tóm tắt (cần ít nhất $_minWords từ)';
    }
    try {
      _cloudConsentStore?.requireConsent();
    } on CloudAnalysisConsentRequiredException {
      return 'Phân tích cloud chưa được đồng ý; giữ kết quả offline.';
    }
    final prompt = PromptBuilder.buildSummarizationPrompt(trimmed);
    final result = await _client.query<String>(
      prompt,
      (responseText, _) => responseText.trim(),
    );
    return result.fold(
      onSuccess: (value) => value,
      onFailure: (error, _) => 'Lỗi tóm tắt: $error',
    );
  }
}
